File: code/ex430.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node prev;
    public Node next;
    public Node child;
};
*/
class Solution {
    public Node flatten(Node head) {
        if (head == null)    return head;
        Node end = dfs(head);
        return head;
    }
    
    private Node dfs(Node head) {
        Node p = head;
        
        if (p.next == null) {
            if (p.child != null) {
                Node end = dfs(p.child);
                p.next = p.child;
                p.next.prev = p;
                p.child = null;
            }
            return p;
        }
        
        while (p.next != null) {
            if (p.child != null) {
                Node nxt = p.next;
                Node end = dfs(p.child); 
                end.next = nxt;
                nxt.prev = end;
                p.next = p.child;
                p.next.prev = p;
                p.child = null;
                p = nxt;
            }
            else
                p = p.next;
        }
        
        return p;
    }
}

File: code/ex430.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* prev;
    Node* next;
    Node* child;
};
*/

class Solution {
public:
    Node* flatten(Node* head) {
        Node* p = head;
        while (p != NULL) {
            if (p->child != NULL) {
                Node* nxt = p->next;
                Node* nxtLevel = flatten(p->child);
                p->next = flatten(p->child);
                p->child = NULL;
                p->next->prev = p;
                while (p->next != NULL)    p = p->next;
                p->next = nxt;
                if (nxt != NULL)    nxt->prev = p;
            }
            p = p->next;
        }
        return head;
    }
};